Tests/External/Plugins/ASCompletion.Tests/Completion/ASGeneratorTests.cs

```diff
@@ -0,0 +1,611 @@
+using System.Collections.Generic;
+using AS3Context;
+using ASCompletion.Context;
+using ASCompletion.Model;
+using ASCompletion.Settings;
+using ASCompletion.TestUtils;
+using FlashDevelop;
+using NSubstitute;
+using NUnit.Framework;
+using PluginCore;
+using ScintillaNet;
+using ScintillaNet.Enums;
+
+namespace ASCompletion.Completion
+{
+    [TestFixture]
+    public class ASGeneratorTests
+    {
+        private MainForm mainForm;
+        private ISettings settings;
+        private ITabbedDocument doc;
+
+        [TestFixtureSetUp]
+        public void FixtureSetUp()
+        {
+            mainForm = new MainForm();
+            settings = Substitute.For<ISettings>();
+            settings.UseTabs = true;
+            settings.IndentSize = 4;
+            settings.SmartIndentType = SmartIndent.CPP;
+            settings.TabIndents = true;
+            settings.TabWidth = 4;
+            doc = Substitute.For<ITabbedDocument>();
+            mainForm.Settings = settings;
+            mainForm.CurrentDocument = doc;
+            mainForm.StandaloneMode = false;
+            PluginBase.Initialize(mainForm);
+            FlashDevelop.Managers.ScintillaManager.LoadConfiguration();
+        }
+
+        [TestFixtureTearDown]
+        public void FixtureTearDown()
+        {
+            settings = null;
+            doc = null;
+            mainForm.Dispose();
+            mainForm = null;
+        }
+
+        private ScintillaControl GetBaseScintillaControl()
+        {
+            return new ScintillaControl
+            {
+                Encoding = System.Text.Encoding.UTF8,
+                CodePage = 65001,
+                Indent = settings.IndentSize,
+                Lexer = 3,
+                StyleBits = 7,
+                IsTabIndents = settings.TabIndents,
+                IsUseTabs = settings.UseTabs,
+                TabWidth = settings.TabWidth
+            };
+        }
+
+        public class GetBodyStart : ASGeneratorTests
+        {
+            [Test]
+            public void SimpleCase()
+            {
+                var sci = GetBaseScintillaControl();
+                sci.Text = "function test():void{\r\n\t\t\t\r\n}";
+                sci.ConfigurationLanguage = "haxe";
+                sci.Colourise(0, -1);
+                int funcBodyStart = ASGenerator.GetBodyStart(0, 1, sci);
+
+                Assert.AreEqual(26, funcBodyStart);
+            }
+
+            [Test]
+            public void EndOnSecondLine()
+            {
+                var sci = GetBaseScintillaControl();
+                // TODO: Should we reindent second line?
+                sci.Text = "function test():void{\r\n\t\t\t}";
+                sci.ConfigurationLanguage = "haxe";
+                sci.Colourise(0, -1);
+                int funcBodyStart = ASGenerator.GetBodyStart(0, 1, sci);
+
+                Assert.AreEqual(26, funcBodyStart);
+                Assert.AreEqual("function test():void{\r\n\t\t\t\r\n}", sci.Text);
```

Comment:
Once again, whenever you see patterns repeating in your tests, try to abstract out the common code and use TestCase attributes to drive the differences. This test might become something like,

```C#
[TestCase("function test():void{\r\n\t\t\t}", "function test():void{\r\n\t\t\t\r\n}", Description = "End on second line")]
[TestCase("function test():void{\r\n}", "function test():void{\r\n\t\r\n}", Description = "End on second line no extra indent")]
public void BodyStartTests(string text, string expected)
{
    var sci = GetBaseScintillaControl();
    sci.Text = text;
    sci.ConfigurationLanguage = "haxe";
    sci.Colourise(0, -1);
    int funcBodyStart = ASGenerator.GetBodyStart(0, 1, sci);
 
    Assert.AreEqual(24, funcBodyStart);
    Assert.AreEqual(expected, sci.Text);
}
```

If there are more differences, add parameters for them.